### PR TITLE
JavaScript: Adds new library functionality: `Function::getAnUndefinedReturn`

### PR DESCRIPTION
**Pull request by psygnisfive**
This method on the `Function` class makes it easy to get places in code where a
function can return undefined via an implicit return value. From the inline
docs:

Functions can sometimes return without returning a value, in which case they
"return" `undefined`. They can do this in two ways:

1. An explicit return statement with no expression, i.e. the statement `return;`

2. An implicit return resulting from an expression executing as the last thing
   in the function. For example, the test in a final `if` statement:

   ```
   function foo() {
     ...
     if (test) { return 1; }
   }
   ```

Some things look like they might return undefined but actually don't because
the containing functioning doesn't return at all. For instance, `throw`
statements prevent the containing function from returning, so they don't count
as undefined returns. Similarly, `yield` doesn't actually cause a return,
since the containing function is a generator and can be re-entered, so we also
exclude yields entirely.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 1]
```diff
@@ -158,6 +158,40 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
     result = getAReturnStmt().getExpr()
   }
 
+  /**
+   * 
+   * Functions can sometimes return without returning a value, in which case they
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Functions.qll]: Please see the [QLDoc Style Guide](https://wiki.semmle.com/display/IN/QLDoc+Style+Guide) (internal link) for guidelines on how to write doc comments; the other doc comments in this file provide concrete examples.

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 2]
```diff
@@ -158,6 +158,40 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
...
+   * exclude yields entirely.
+   */
+  
+  ConcreteControlFlowNode getAnUndefinedReturn() {
+    not (this instanceof ArrowFunctionExpr and this.getBody() instanceof Expr) and
+    result.getContainer() = this and
+    result.isAFinalNode() and
+    not (result instanceof ReturnStmt and exists(result.(ReturnStmt).getExpr())) and
+    not result instanceof ThrowStmt and
+    not result instanceof YieldExpr
```
[the remark this answers is not in the document]
Comment (psygnisfive) [reply javascript/ql/src/semmle/javascript/Functions.qll]: Do you think it would make sense instead to ignore these functions, then, or would it be useful to try to capture them too?

Given the use cases of generator functions and async functions, it might be a good approximation to simply let those functions do what they want, b/c they're not intended to be used for their returns?

File: javascript/ql/test/library-tests/Functions/tests.expected [hunk 3]
```diff
@@ -168,3 +168,22 @@ test_isRestParameter
 | restparms.js:1:18:1:19 | ys |
 test_ReturnStmt
 | tst.js:14:1:14:37 | functio ... s[0]; } | tst.js:14:16:14:35 | return arguments[0]; |
+test_getAnUndefinedReturn
```
[the remark this answers is not in the document]
Comment (psygnisfive) [reply javascript/ql/test/library-tests/Functions/tests.expected]: I'm uncertain what you mean here; can you clarify?

File: javascript/ql/src/semmle/javascript/Functions.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Functions.qll]: Yes, just ignore generators and `async` functions.

File: javascript/ql/test/library-tests/Functions/tests.expected (hunk 3 above)
[the remark this answers is not in the document]
Comment (xiemaisi) [reply javascript/ql/test/library-tests/Functions/tests.expected]: Never mind, I see that you didn't actually add any new test files. Do the existing files exercise all the various corner cases of your new predicate? For instance, I don't see any `async` functions in the existing test files.

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 4]
```diff
@@ -158,6 +158,45 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
     result = getAReturnStmt().getExpr()
   }
 
+  /** 
+   * Gets a `ConcreteControlFlowNode` which corresponds to an executable portion of
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Functions.qll]: Could you try simplifying this a little? I realise you want it to be technically accurate, but one of the most important uses of these comments is that they are shown during autocomplete, and for that purpose it's better if the first sentence is short and easy to grasp intuitively.

You can then add a more precise technical definition as a second (third, ...) paragraph, which won't be shown as part of the autocomplete hover.

File: javascript/ql/test/library-tests/Functions/tests.expected (hunk 3 above)
[the remark this answers is not in the document]
Comment (psygnisfive) [reply javascript/ql/test/library-tests/Functions/tests.expected]: I added `getAnUndefinedReturn.qll` but no new `.js` files. I'll add one specifically for this functionality, one that specifically targets all of the cases.

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 5]
```diff
@@ -158,6 +158,40 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
...
+   * Some things look like they might return undefined but actually don't because
+   * the containing functioning doesn't return at all. For instance, `throw`
+   * statements prevent the containing function from returning, so they don't count
+   * as undefined returns. Similarly, `yield` doesn't actually cause a return,
+   * since the containing function is a generator and can be re-entered, so we also
+   * exclude yields entirely.
+   */
+  
+  ConcreteControlFlowNode getAnUndefinedReturn() {
+    not (this instanceof ArrowFunctionExpr and this.getBody() instanceof Expr) and
```
[the remark this answers is not in the document]
Comment (psygnisfive) [reply javascript/ql/src/semmle/javascript/Functions.qll]: Regarding this, it seems like the intention behind your proposed change is to make it so that functions such as `function() 1` will be caught, and that this therefore requires the `semmle-extractor-options: --experimental` flag in the relevant test `.js` file. I've included this now, tho if this was not the intention, please let me know.

[the remark this answers is not in the document]
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Functions.qll]: That's exactly what I meant. :+1:

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 6]
```diff
@@ -158,6 +158,51 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
     result = getAReturnStmt().getExpr()
   }
 
+  /** 
+   * Gets a return from a function which has undefined value (i.e. implicit returns
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Functions.qll]: Minor: we don't normally use "i.e." and "e.g." in doc comments.

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 7]
```diff
@@ -158,6 +158,51 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
...
+   * and returns w/o expressions).
+   */
+   
+  /* Functions can have undefined returns in a few different ways:
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Functions.qll]: You can just join this comment with the previous one. Generally, the first paragraph of a doc comment gives an overview (as the one above does), while the rest of the comment provides technical details.

File: javascript/ql/src/semmle/javascript/Functions.qll [hunk 8]
```diff
@@ -158,6 +158,51 @@ class Function extends @function, Parameterized, TypeParameterized, StmtContaine
     result = getAReturnStmt().getExpr()
   }
 
+  /** 
+   * Gets a return from a function which has undefined value (i.e. implicit returns
+   * and returns w/o expressions).
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/Functions.qll]: No informal or non-standard abbreviations, please, they can be difficult to understand for us non-native speakers.

File: javascript/ql/test/library-tests/Functions/undefinedreturns.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+async function async_fn() { }
+function fn_w_final_throw() { throw 1; }
+function* fn_w_final_yield() { yield 1; }
+function fn_w_final_return_w_expr() { return 1; }
+
+
+
+/////////////////
+//             //
+//  BAD CASES  //
```
Comment (xiemaisi) [inline javascript/ql/test/library-tests/Functions/undefinedreturns.js]: I don't think "good"/"bad" labels apply in this case, since you're not testing a query, just a library predicate.
